File: src/2642.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <queue>
#include <vector>

using namespace std;

class Graph {
   public:
    vector<vector<pair<int, int>>> adjList;
    Graph(int n, vector<vector<int>>& edges) {
        adjList.resize(n);
        for (auto& e : edges) {
            adjList[e[0]].push_back(make_pair(e[1], e[2]));
        }
    }
    void addEdge(vector<int> edge) {
        //
        adjList[edge[0]].push_back(make_pair(edge[1], edge[2]));
    }
    int shortestPath(int node1, int node2) {
        int retVal = -1;

        int n = adjList.size();
        vector<int> costForNode(n, numeric_limits<int>::max());
        costForNode[node1] = 0;

        priority_queue<vector<int>, vector<vector<int>>, greater<vector<int>>> pq;
        pq.push({0, node1});
        while (pq.empty() == false) {
            int currCost = pq.top()[0];
            int currNode = pq.top()[1];
            pq.pop();
            if (currCost > costForNode[currNode]) {
                continue;
            }

            if (currNode == node2) {
                retVal = currCost;
                break;
            }

            for (auto& neighbor : adjList[currNode]) {
                int neighborNode = neighbor.first;
                int cost = neighbor.second;
                int newCost = currCost + cost;
                if (newCost < costForNode[neighborNode]) {
                    costForNode[neighborNode] = newCost;
                    pq.push({newCost, neighborNode});
                }
            }
        }

        return retVal;
    }
};
/**
 * Your Graph object will be instantiated and called as such:
 * Graph* obj = new Graph(n, edges);
 * obj->addEdge(edge);
 * int param_2 = obj->shortestPath(node1,node2);
 */

int main(int argc, char** argv) {
    struct subject {
        vector<string> methods;
        vector<vector<vector<int>>> edges;
        vector<vector<int>> edge;
        vector<vector<int>> node;
    };
    vector<subject> testData{{{"Graph", "shortestPath", "shortestPath", "addEdge", "shortestPath"},
                              {{{{0, 2, 5}, {0, 1, 2}, {1, 2, 1}, {3, 0, 3}}}, {{}}, {{}}, {{}}, {{}}},
                              {{}, {}, {}, {1, 3, 4}, {}},
                              {{}, {3, 2}, {0, 3}, {}, {0, 3}}}};
    int numberOfTestCase = testData.size();
    /* Example
     *  Input
     *  ["Graph", "shortestPath", "shortestPath", "addEdge", "shortestPath"]
     *  [[4, [[0, 2, 5], [0, 1, 2], [1, 2, 1], [3, 0, 3]]], [3, 2], [0, 3], [[1, 3, 4]], [0, 3]]
     *  Output
     *  [null, 6, -1, null, 6]
     */

    for (int i = 0; i < numberOfTestCase; ++i) {
        cout << "Input\n";
        cout << "[";
        for (long unsigned int j = 0; j < testData[i].methods.size(); ++j) {
            cout << (j == 0 ? "" : ", ") << '"' << testData[i].methods[j] << '"';
        }
        cout << "]\n";
        cout << "[";
        for (long unsigned int j = 0; j < testData[i].methods.size(); ++j) {
            cout << (j == 0 ? "" : ", ") << "[";
            if (testData[i].methods[j].compare("Graph") == 0) {
                cout << testData[i].edges[j].size() << ", [";
                for (long unsigned int k = 0; k < testData[i].edges[j].size(); ++k) {
                    cout << (k == 0 ? "" : ", ") << "[";
                    for (long unsigned int l = 0; l < testData[i].edges[j][k].size(); ++l) {
                        cout << (l == 0 ? "" : ", ") << testData[i].edges[j][k][l];
                    }
                    cout << "]";
                }
                cout << "]";
            } else if (testData[i].methods[j].compare("shortestPath") == 0) {
                for (long unsigned int k = 0; k < testData[i].node[j].size(); ++k) {
                    cout << (k == 0 ? "" : ", ") << testData[i].node[j][k];
                }
            } else if (testData[i].methods[j].compare("addEdge") == 0) {
                cout << "[";
                for (long unsigned int k = 0; k < testData[i].edge[j].size(); ++k) {
                    cout << (k == 0 ? "" : ", ") << testData[i].edge[j][k];
                }
                cout << "]";
            }
            cout << "]";
        }
        cout << "]\n";

        Graph cGraph(testData[i].edges[0].size(), testData[i].edges[0]);
        cout << "Output\n";
        cout << "[";
        for (long unsigned int j = 0; j < testData[i].methods.size(); ++j) {
            if (testData[i].methods[j].compare("Graph") == 0) {
                cout << (j == 0 ? "" : ", ") << "null";
            } else if (testData[i].methods[j].compare("addEdge") == 0) {
                cGraph.addEdge(testData[i].edge[j]);
                cout << (j == 0 ? "" : ", ") << "null";
            } else if (testData[i].methods[j].compare("shortestPath") == 0) {
                int retVal = cGraph.shortestPath(testData[i].node[j][0], testData[i].node[j][1]);
                cout << (j == 0 ? "" : ", ") << retVal;
            }
        }
        cout << "]\n";

        cout << "\n";
    }

    return EXIT_SUCCESS;
}

File: src/2642.py
import sys
import logging
from typing import List, Optional
from collections import defaultdict, deque, Counter
from functools import cache, lru_cache
from bisect import bisect_left, bisect_right
from itertools import accumulate
from heapq import heapify, heappush, heappop


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Graph:
    def __init__(self, n: int, edges: List[List[int]]):
        self.adj_list = [[] for _ in range(n)]
        for from_node, to_node, cost in edges:
            self.adj_list[from_node].append((to_node, cost))

    def addEdge(self, edge: List[int]) -> None:
        from_node, to_node, cost = edge
        self.adj_list[from_node].append((to_node, cost))

    def shortestPath(self, node1: int, node2: int) -> int:
        retVal = -1

        n = len(self.adj_list)
        cost_for_node = [float('inf')] * (n)
        cost_for_node[node1] = 0

        pq = [(0, node1)]
        while pq:
            curr_cost, curr_node = heappop(pq)
            if curr_cost > cost_for_node[curr_node]:
                continue

            if curr_node == node2:
                retVal = curr_cost
                break

            for neighbor, cost in self.adj_list[curr_node]:
                new_cost = curr_cost + cost
                if new_cost < cost_for_node[neighbor]:
                    cost_for_node[neighbor] = new_cost
                    heappush(pq, (new_cost, neighbor))

        return retVal


# Your Graph object will be instantiated and called as such:
# obj = Graph(n, edges)
# obj.addEdge(edge)
# param_2 = obj.shortestPath(node1,node2)

if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        logging.info("Input")
        for methods, graphType in zip(
            [["Graph", "shortestPath", "shortestPath", "addEdge", "shortestPath"]],
                [[[4, [[0, 2, 5], [0, 1, 2], [1, 2, 1], [3, 0, 3]]], [3, 2], [0, 3], [[1, 3, 4]], [0, 3]]]):
            # /* Example
            #  *  Input
            #  *  ["Graph", "shortestPath", "shortestPath", "addEdge", "shortestPath"]
            #  *  [[4, [[0, 2, 5], [0, 1, 2], [1, 2, 1], [3, 0, 3]]], [3, 2], [0, 3], [[1, 3, 4]], [0, 3]]
            #  *  Output
            #  *  [null, 6, -1, null, 6]
            #  */
            logging.info("%s", methods)
            logging.info("%s", graphType)

            displayStr = []
            for key, value in zip(methods, graphType):
                if key == "Graph":
                    obj = Graph(value[0], value[1])
                    displayStr.append(None)
                elif key == "addEdge":
                    obj.addEdge(value[0])
                    displayStr.append(None)
                elif key == "shortestPath":
                    retVal = obj.shortestPath(value[0], value[1])
                    displayStr.append(retVal)
            logging.info("Output")
            logging.info("%s", displayStr)

            print("")
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__, exception, exc_info=True)
        pass

    sys.exit(0)
